File: tests/wasm/test_wma.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('WMA partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.wma_js(close, 30);
    assert.strictEqual(result.length, close.length);
});

test('WMA accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.wma;

    const result = wasm.wma_js(
        close,
        expected.defaultParams.period
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-6,
        "WMA last 5 values mismatch"
    );


    await compareWithRust('wma', result, 'close', expected.defaultParams);
});

test('WMA default candles', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.wma_js(close, 30);
    assert.strictEqual(result.length, close.length);
});

test('WMA empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.wma_js(empty, 30);
    }, /Input data slice is empty/);
});

test('WMA zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.wma_js(inputData, 0);
    }, /Invalid period/);
});

test('WMA period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.wma_js(dataSmall, 10);
    }, /Invalid period/);
});

test('WMA very small dataset', () => {

    const singlePoint = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.wma_js(singlePoint, 9);
    }, /Invalid period|Not enough valid data/);
});

test('WMA reinput', () => {

    const close = new Float64Array(testData.close);


    const firstResult = wasm.wma_js(close, 14);
    assert.strictEqual(firstResult.length, close.length);


    const secondResult = wasm.wma_js(firstResult, 5);
    assert.strictEqual(secondResult.length, firstResult.length);


    if (secondResult.length > 50) {
        for (let i = 50; i < secondResult.length; i++) {
            assert(!isNaN(secondResult[i]), `Unexpected NaN at index ${i}`);
        }
    }
});

test('WMA NaN handling', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.wma_js(close, 14);
    assert.strictEqual(result.length, close.length);


    if (result.length > 50) {
        for (let i = 50; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }




    assertAllNaN(result.slice(0, 13), "Expected NaN in warmup period (indices 0-12)");
    assert(!isNaN(result[13]), "Expected first valid value at index 13");
});

test('WMA all NaN input', () => {

    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);

    assert.throws(() => {
        wasm.wma_js(allNaN, 30);
    }, /All values are NaN/);
});

test('WMA batch single parameter set', () => {

    const close = new Float64Array(testData.close);


    const batchResult = wasm.wma_batch(close, {
        period_range: [30, 30, 0]
    });


    const singleResult = wasm.wma_js(close, 30);

    assert.strictEqual(batchResult.values.length, singleResult.length);
    assert.strictEqual(batchResult.rows, 1);
    assert.strictEqual(batchResult.cols, close.length);
    assert.strictEqual(batchResult.combos.length, 1);
    assert.strictEqual(batchResult.combos[0].period, 30);


    const p = 30;
    const firstValid = p - 1;

    assertClose(batchResult.values[firstValid], singleResult[firstValid], 1e-6, "First valid mismatch");

    const end = batchResult.values.length;
    assertArrayClose(batchResult.values.slice(end - 5, end), singleResult.slice(end - 5, end), 1e-6, "Last 5 mismatch");


});

test('WMA batch multiple periods', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const batchResult = wasm.wma_batch(close, {
        period_range: [10, 30, 10]
    });


    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    assert.strictEqual(batchResult.combos.length, 3);


    const periods = [10, 20, 30];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.values.slice(rowStart, rowEnd);

        const singleResult = wasm.wma_js(close, periods[i]);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch`
        );


        assert.strictEqual(batchResult.combos[i].period, periods[i]);
    }
});

test('WMA batch metadata', () => {

    const metadata = wasm.wma_batch_metadata_js(
        10, 30, 10
    );


    assert.strictEqual(metadata.length, 3);


    assert.strictEqual(metadata[0], 10);
    assert.strictEqual(metadata[1], 20);
    assert.strictEqual(metadata[2], 30);
});

test('WMA batch full parameter sweep', () => {

    const close = new Float64Array(testData.close.slice(0, 50));

    const batchResult = wasm.wma_batch_js(
        close,
        10, 14, 2
    );

    const metadata = wasm.wma_batch_metadata_js(
        10, 14, 2
    );


    const numCombos = metadata.length;
    assert.strictEqual(numCombos, 3);
    assert.strictEqual(batchResult.length, 3 * 50);


    for (let combo = 0; combo < numCombos; combo++) {
        const period = metadata[combo];

        const rowStart = combo * 50;
        const rowData = batchResult.slice(rowStart, rowStart + 50);


        for (let i = 0; i < period - 1; i++) {
            assert(isNaN(rowData[i]), `Expected NaN at warmup index ${i} for period ${period}`);
        }


        for (let i = period - 1; i < 50; i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for period ${period}`);
        }
    }
});

test('WMA batch edge cases', () => {

    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);


    const singleBatch = wasm.wma_batch_js(
        close,
        5, 5, 1
    );

    assert.strictEqual(singleBatch.length, 10);


    const largeBatch = wasm.wma_batch_js(
        close,
        5, 7, 10
    );


    assert.strictEqual(largeBatch.length, 10);
});



test.after(() => {
    console.log('WMA WASM tests completed');
});

File: tests/python/test_wma.py
"""
Python binding tests for WMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

# Add parent directory to path to import the built module
sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'target/wheels'))

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestWma:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_wma_partial_params(self, test_data):
        """Test WMA with partial parameters (None values) - mirrors check_wma_partial_params"""
        close = test_data['close']
        
        # Test with default params (period=30)
        result = ta_indicators.wma(close, 30)  # Using default
        assert len(result) == len(close)
    
    def test_wma_accuracy(self, test_data):
        """Test WMA matches expected values from Rust tests - mirrors check_wma_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['wma']
        
        result = ta_indicators.wma(
            close,
            period=expected['default_params']['period']
        )
        
        assert len(result) == len(close)
        
        # Check last 5 values match expected
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=1e-6,  # Using 1e-6 as per Rust test
            msg="WMA last 5 values mismatch"
        )
    
    def test_wma_zero_period(self):
        """Test WMA fails with zero period - mirrors check_wma_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.wma(input_data, period=0)
    
    def test_wma_period_exceeds_length(self):
        """Test WMA fails when period exceeds data length - mirrors check_wma_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.wma(data_small, period=10)
    
    def test_wma_very_small_dataset(self):
        """Test WMA fails with insufficient data - mirrors check_wma_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.wma(single_point, period=9)
    
    def test_wma_reinput(self, test_data):
        """Test WMA applied twice (re-input) - mirrors check_wma_reinput"""
        close = test_data['close']
        
        # First pass
        first_result = ta_indicators.wma(close, period=14)
        assert len(first_result) == len(close)
        
        # Second pass - apply WMA to WMA output
        second_result = ta_indicators.wma(first_result, period=5)
        assert len(second_result) == len(first_result)
        
        # Check that values after warmup are not NaN
        if len(second_result) > 50:
            for i in range(50, len(second_result)):
                assert not np.isnan(second_result[i])
    
    def test_wma_nan_handling(self, test_data):
        """Test WMA handles NaN values correctly - mirrors check_wma_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.wma(close, period=14)
        assert len(result) == len(close)
        
        # After warmup period, no NaN values should exist
        if len(result) > 50:
            assert not np.any(np.isnan(result[50:])), "Found unexpected NaN after warmup period"
        
        # First period-1 values should be NaN (warmup is period-1 for WMA)
        # For period=14, indices 0-12 should be NaN
        assert np.all(np.isnan(result[:13])), "Expected NaN in warmup period"
        # First valid value should be at index period-1 = 13
        assert not np.isnan(result[13]), "Expected valid value at index 13"
    
    def test_wma_streaming(self, test_data):
        """Test WMA streaming matches batch calculation - mirrors check_wma_streaming"""
        close = test_data['close']
        period = 30
        
        # Batch calculation
        batch_result = ta_indicators.wma(close, period=period)
        
        # Streaming calculation
        stream = ta_indicators.WmaStream(period=period)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        # Compare batch vs streaming
        assert len(batch_result) == len(stream_values)
        
        # Compare values where both are not NaN
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-8, atol=1e-8, 
                        msg=f"WMA streaming mismatch at index {i}")
    
    def test_wma_batch(self, test_data):
        """Test WMA batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.wma_batch(
            close,
            period_range=(30, 30, 0)  # Default period only
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        # Should have 1 combination (default params)
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        # Extract the single row
        default_row = result['values'][0]
        expected = EXPECTED_OUTPUTS['wma']['last_5_values']
        
        # Check last 5 values match
        assert_close(
            default_row[-5:],
            expected,
            rtol=1e-6,
            msg="WMA batch default row mismatch"
        )
    
    def test_wma_all_nan_input(self):
        """Test WMA with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.wma(all_nan, period=30)
    
    def test_wma_kernel_selection(self, test_data):
        """Test WMA with different kernel selections"""
        close = test_data['close'][:100]  # Use smaller dataset for speed
        
        # Test different kernels
        kernels = ['auto', 'scalar', 'avx2', 'avx512']
        results = {}
        
        for kernel in kernels:
            try:
                results[kernel] = ta_indicators.wma(
                    close, 
                    period=30,
                    kernel=kernel
                )
            except ValueError as e:
                # Some kernels might not be available on all systems
                if "Unknown kernel" not in str(e) and "not available on this CPU" not in str(e) and "not compiled in this build" not in str(e):
                    raise
        
        # All available kernels should produce similar results
        if 'scalar' in results:
            for kernel, result in results.items():
                if kernel != 'scalar':
                    assert_close(
                        result,
                        results['scalar'],
                        rtol=1e-10,
                        msg=f"Kernel {kernel} mismatch with scalar"
                    )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
